### ISSUE TITLE
Do not support MultiGPU training

### ISSUE THREAD

**Issue by gang-wu-yi** [david8862/keras-YOLOv3-model-set#56]
hi,guys. Thank you for your coding, but the repo is not support multi_gpus training. Although this repo contains multi_gpus code , it is no work well.  When i training model by 2_gpus , two gpus is occupied, but only one is working or computing, another one is idle.
so, do you wanna to fix it? or how to fix it ?

**Comment by david8862** [comment]
Hi @gang-wu-yi , currently multi-gpu support in this repo still leverage the old multi_gpu_model() function in keras from [keras-yolo3](https://github.com/qqwweee/keras-yolo3). I didn't change it since don't have multi-gpu environment to test new code :). Actually TF 2.x have provided new [distributed_training](https://www.tensorflow.org/guide/distributed_training) API to support multi-gpu. I'll try to integrated it here later.

**Comment by david8862** [comment]
Hi @gang-wu-yi , I've tried to improve MultiGPU training with tf.distribute.MirroredStrategy in latest code, but not verified. You can have a try to see if the fix can help :)

**Comment by gang-wu-yi** [comment]
wow, @david8862 , thank you for your reply.
 i will verify your MultiGPU training code , and reply to you later.
and , during train model , i have some problem as follows:
1. model loss / total loss is right, means reduce normally, but accuracy is 0 always , i checked annotation and exchange optimizer(from adam to sgd), accuracy is still 0 , do you have any advice?
2. about kmeans.py , your kmeans.py is different from others(ex. qqwweee/keras-yolo3/kmeans.py), i generate anchores by  qqwweee/keras-yolo3/kmeans.py, droped your kmeans.py , i don't know how to use your kmeans, and generate anchores from your code base on 608*608 image, it looks wrong, right??
3. monitor learning rate on tensorboard
as i set --decay_type='cosine' and remove callbacks.remove(reduce_lr) from train.py , some error happens


Really thanks for your great implementation of yolo mode....

**Comment by david8862** [comment]
@gang-wu-yi 

1. what do you mean by "accuracy"? For detection models a common metric is PascalVOC mAP or COCO AP. You can online or offline evaluate your trained model with "--online_eval" option or eval.py
2. kmeans.py here is just inherited from qqwweee's version, but adding some CLI options for usage convenience. Generally you could choose anyone to generate your custom anchors.
3. that's an incorrect modification. reduce_lr callback shouldn't be used together with any decay_type

**Comment by gang-wu-yi** [comment]
@david8862 , thank you for great implementation of yolo mode
About MultiGPU training as follows:
1. I had trained two times , first time is ok , means training model via mini epochs(15) , 20000 images , The training process is normal，and trained weight or final weight is able to predict, but the prediction is not perfect , so i said it's ok, logs as as bellow:
![image](https://user-images.githubusercontent.com/50908195/85965895-c9f83d00-b9f0-11ea-9d72-bf932197c8f6.png)
2. But second time , some issue happens :
    - take more epochs , but loss is higher than first time; 
    - make prediction is wrong, means draw box is wrong obviously ，such as some boxes draw on one object ... , maybe i do some mistakes on it ; 
![image](https://user-images.githubusercontent.com/50908195/85966511-96b6ad80-b9f2-11ea-8e93-baf79406102b.png)
3. Notice ! Notice ! it is sleep or hang up or time-out that during training process, either 1_gpu or multi_gpus , sometime it can be resume by close and reopen pycharm, sometimes it faid to resume , it is weird, i had checked my GPU usage--not stable, i I didn't find the reason .

**Comment by david8862** [comment]
Hi @gang-wu-yi 
2. model.compile() for 1st stage has been done in get_train_model() so I didn't repeat it in the main process.
3. Regarding the hang up issue, I didn't meet that or get similar feedback. Have you ever tried to run on bare console instead of pycharm?

**Comment by gang-wu-yi** [comment]
Hi @david8862 ,  and some guys , please know as follow:
1. The repo do support MultiGPUs during training, and I've verified it before two weeks. 
2. Regarding the hang up issue -- I disconnect pycharm from the GPU's server during training , the issue disappeared, i don't know why , but i can duplicate it completely when i do training under keep-alive status( means pycharm client and GPUs server keep connect ,  to show training logs on pycharm client).

**Comment by david8862** [comment]
@gang-wu-yi  many thanks. And currently the model construction doesn't support Sync BN, which will impact MultiGPU training. I'll add it into TODO list and schedule to implement.
